File: src/main/java/nz/co/searchwellington/feeds/FeedItemAcceptor.java
package nz.co.searchwellington.feeds;

import nz.co.searchwellington.model.Newsitem;
import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.utils.UrlFilters;

import org.apache.log4j.Logger;
import org.joda.time.DateTime;
import org.springframework.stereotype.Component;

@Component
public class FeedItemAcceptor {
	
	private static Logger log = Logger.getLogger(FeedItemAcceptor.class);
	
	public Newsitem acceptFeedItem(User user, Newsitem newsitem) {
		log.info("Accepting: " + newsitem.getName() + " from " + newsitem.getFeed().getName());
		flattenLoudCapsInTitle(newsitem);     
        if (newsitem.getDate() == null) {
        	log.info("Accepting a feeditem with no date; setting date to current time");            
        	newsitem.setDate(new DateTime().toDate());
        }
        
        newsitem.setAccepted(new DateTime().toDate());
        newsitem.setAcceptedBy(user);
        newsitem.setOwner(user);
        return newsitem;
    }
	
    // TODO Push to service for testing.
	private void flattenLoudCapsInTitle(Resource resource) {		
		String flattenedTitle = UrlFilters.lowerCappedSentence(resource.getName());           
        if (!flattenedTitle.equals(resource.getName())) {
        	resource.setName(flattenedTitle);
            log.info("Flatten capitalised sentence to '" + flattenedTitle + "'");
        }
	}
	
}

File: src/main/scala/nz/co/searchwellington/feeds/FeedItemAcceptor.scala
package nz.co.searchwellington.feeds

import nz.co.searchwellington.feeds.whakaoko.model.FeedItem
import nz.co.searchwellington.model.{Feed, FeedAcceptancePolicy, Newsitem, User}
import org.joda.time.DateTime
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component class FeedItemAcceptor @Autowired()(feeditemToNewsItemService: FeeditemToNewsitemService) {

  def acceptFeedItem(acceptingUser: User, feedItemAndFeed: (FeedItem, Feed)): Option[Newsitem] = {
    // Given a feed item attempt to transform it into an acceptable news item
    val feedItem = feedItemAndFeed._1
    val feed = feedItemAndFeed._2

    feeditemToNewsItemService.makeNewsitemFromFeedItem(feedItem, feed).map { newsitem =>
      // Apply acceptance details to the news item
      val now = DateTime.now.toDate
      val dateToApply = feed.acceptance match {
        case FeedAcceptancePolicy.ACCEPT_IGNORING_DATE => now
        case _ => newsitem.date.getOrElse(feedItem.accepted.map(_.toDate).getOrElse(now))
      }
      newsitem.copy(
        date = Some(dateToApply),
        accepted = Some(now),
        acceptedBy = Some(acceptingUser._id),
        owner = Some(acceptingUser._id)
      )
    }
  }

}
